Scene improvements (#1651)

gran4 opened the pull request:
attempt to fix https://github.com/pythonarcade/arcade/issues/1651

gran4 committed:
Fixes collisions. 

Before, if min_a, min_b, max_a, max_b weren't changed by the code, it would break(make odd collisions) as https://github.com/pythonarcade/arcade/issues/1183 notes. Now, it always creates the collisions

gran4 committed:
Update geometry.py

gran4 committed:
Merge branch 'pythonarcade:development' into development

gran4 committed:
Update scene.py

gran4 committed:
Update scene.py

gran4 committed:
Update scene.py

remove duplicate sprite lists

gran4 committed:
Update scene.py

made vars non-public

gran4 committed:
Update scene.py

added warn library, lol

gran4 commented:
I think the warnings I added are breaking the tests, it says "Warning, treated as error:".

einarf commented on arcade/scene.py:
```diff
@@ -20,17 +21,18 @@ class Scene:
-        :sprite_lists: A list of `SpriteList` objects. The order of this list is the
+        :DON'T CHANGE __sprite_lists or __name_mapping
+        :__sprite_lists: A list of `SpriteList` objects. The order of this list is the
                        order in which they will be drawn.
-        :name_mapping: A dictionary of `SpriteList` objects. This contains the same lists
-                       as the `sprite_lists` attribute, but is a mapping of them by name. This is
-                       not necessarily in the same order as the `sprite_lists` attribute.
+        :__name_mapping: A dictionary of `SpriteList` objects. This contains the same lists
+                       as the `__sprite_lists` attribute, but is a mapping of them by name. This is
+                       not necessarily in the same order as the `__sprite_lists` attribute.(Dictionaries have no order)
```
If these are non-public this entire section should just be removed.

gran4 replied:
that's true

gran4 replied:
Fixed. @einarf.
Thanks

einarf commented on arcade/scene.py:
```diff
@@ -20,17 +21,18 @@ class Scene:
-        self.sprite_lists: List[SpriteList] = []
-        self.name_mapping: Dict[str, SpriteList] = {}
-
+        self.__sprite_lists: List[SpriteList] = []
+        self.__name_mapping: Dict[str, SpriteList] = {}
```
We use single underscore for "private" members.

gran4 replied:
Fixed

gran4 committed:
Update scene.py

einarf commented on arcade/scene.py:
```diff
@@ -20,17 +21,13 @@ class Scene:
-                       order in which they will be drawn.
-        :name_mapping: A dictionary of `SpriteList` objects. This contains the same lists
-                       as the `sprite_lists` attribute, but is a mapping of them by name. This is
-                       not necessarily in the same order as the `sprite_lists` attribute.
+        :DON'T CHANGE _sprite_lists or _name_mapping
```
Just remove this whole attribute section. It's already implied that user should not access these members unless they know exactly what they are doing.

gran4 replied:
Fixed

einarf commented on arcade/scene.py:
```diff
@@ -113,8 +110,11 @@ def add_sprite_list(
             sprite_list = SpriteList(use_spatial_hash=use_spatial_hash)
-        self.name_mapping[name] = sprite_list
-        self.sprite_lists.append(sprite_list)
+        if name in self._name_mapping.keys():
+            self.remove_sprite_list_by_name(name)
+            warn("A Spritelist with the name: "+name+", is already in the scene, will override Spritelist")
```
It's usually better to have a hard error than doing a warning. I'm not sure if this should be a warning/error. Might need some more input from user users/devs. @Cleptomania have any opinion on this?

gran4 replied:
There might be a case where the user doesn't care.

einarf commented on arcade/scene.py:
```diff
@@ -138,10 +138,13 @@ def add_sprite_list_before(
-        before_list = self.name_mapping[before]
-        index = self.sprite_lists.index(before_list)
-        self.sprite_lists.insert(index, sprite_list)
+        if name in self._name_mapping.keys():
+            self.remove_sprite_list_by_name(name)
+            warn("A Spritelist with the name: "+name+", is already in the scene, will override Spritelist")
```
Again: We need to figure out if this should be allowed or not.

einarf commented on arcade/scene.py:
```diff
@@ -190,10 +193,13 @@ def add_sprite_list_after(
-        after_list = self.name_mapping[after]
-        index = self.sprite_lists.index(after_list) + 1
-        self.sprite_lists.insert(index, sprite_list)
+        if name in self._name_mapping.keys():
+            self.remove_sprite_list_by_name(name)
+            warn("A Spritelist with the name: "+name+", is already in the scene, will override Spritelist")
```
Again: We need to figure out if this should be allowed or not.

gran4 committed:
Update text.py

I think it the new create_text_texture should fix https://github.com/pythonarcade/arcade/issues/1410.

einarf commented on arcade/scene.py:
```diff
@@ -307,7 +313,7 @@ def draw(self, names: Optional[List[str]] = None, **kwargs) -> None:
         If `names` parameter is provided then only the specified SpriteLists
         will be drawn. They will be drawn in the order that the names in the
         list were arranged. If `names` is not provided, then every SpriteList
-        in the scene will be drawn according the order of the main sprite_lists
+        in the scene will be drawn according the order of the main _sprite_lists
```
Should probably be phrased wiout mentioning the internal member.

gran4 replied:
Change "the order of the main sprite_lists" to "the order of the sprite lists". Or just delete that? Or something else?

gran4 commented:
oops. I meant to wait. Is there a way to pull it back?

einarf commented:
You can just keep pushing changes. We will squash everything into one commit when merging, so it doesn't matter if it's messy commit history in the PR.

gran4 commented:
I meant to fix another issue(https://github.com/pythonarcade/arcade/issues/1410)

gran4 committed:
Update scene.py

Einaf's Request.

einarf commented:
> I meant to fix another issue(#1410)

`git revert 231541f287c25e209f96ad015289ecaee92019d0` maybe?

gran4 committed:
Update text.py

Revert my mistake change. I meant to make a another pull request, but I guess I'll have to do that later.

gran4 commented:
I just added another commit that reverts my changes.
